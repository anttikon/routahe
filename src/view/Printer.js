import {getColorByMode, getEmojiByMode, formatTime, formatDuration} from './viewUtils'
import {cyan, gray} from 'chalk'

const dateTimeFormat = 'H:mm DD.MM.YYYY'

class Printer {

  addDate(date) {
    if (!date) {
      return this
    }
    return this._addLine(formatTime(date))
  }

  addDuration(duration) {
    if (!duration) {
      return this
    }
    return this._addLine(formatDuration(duration))
  }

  addMode(mode) {
    if (!mode) {
      return this
    }
    const color = getColorByMode(mode)
    return this._addLine(`${getEmojiByMode(mode)} ${color(mode)}`)
  }

  addArrive(arrive) {
    if (!arrive) {
      return this
    }
    this.addString('arrive ', gray)
    return this.addString(arrive.format(dateTimeFormat), cyan)
  }

  addDeparture(departure) {
    if (!departure) {
      return this
    }
    this.addString('departure ', gray)
    return this.addString(departure.format(dateTimeFormat), cyan)
  }

  addString(str, color) {
    if (!str) {
      return this
    }

    return color ? this._addLine(color(str)) : this._addLine(str)
  }

  _addLine(line) {
    if (!this.lines) {
      this.lines = []
    }
    this.lines.push(line)
    return this
  }

  print() {
    this.lines ? console.log(this.lines.join('')) : null
  }
}
export default Printer