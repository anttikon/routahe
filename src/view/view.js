import Printer from './Printer'
import {bold} from 'chalk'
import {head, last, padEnd, get, max} from 'lodash'
import {getColorByMode} from './viewUtils'

exports.printRouteInformation = (route) => {
  new Printer()
    .addDate(head(route.legs).startTime)
    .addString(' - ')
    .addDate(last(route.legs).endTime)
    .addString(' ')
    .addDuration(route.duration)
    .print()
}

exports.printSearchInfo = (from, to, date, arriveBy) => {
  const printer = new Printer()
  printer.addString(from.label, bold)
    .addString(' - ')
    .addString(to.label, bold)

  if (date) {
    printer.addString(' | ')
    arriveBy ? printer.addArrive(date) : printer.addDeparture(date)
  }

  printer.print()
}

exports.printLeg = (route, leg) => {
  const color = getColorByMode(leg.mode)
  const shortName = get(leg, 'route.shortName'),
    padAmount = getShortNamePadAmount(route, leg),
    shortNameWithPadding = padEnd(shortName, padAmount, ' ')

  new Printer()
    .addString('  |  ', color)
    .addDate(leg.startTime)
    .addString(' - ')
    .addDate(leg.endTime)
    .addString(' ')
    .addMode(leg.mode)
    .addString(' ')
    .addString(shortNameWithPadding, bold)
    .addString(' -> ')
    .addString(leg.to.name)
    .print()
}

function getShortNamePadAmount(route, leg) {
  const {mode} = leg
  const longestModeAndShortnameLength = max(route.legs.map(leg => getModeAndShortnameLength(leg)))
  return longestModeAndShortnameLength - mode.length
}

function getModeAndShortnameLength(leg) {
  const {mode} = leg
  const shortName = get(leg, 'route.shortName') || ''
  return mode.length + shortName.length
}

exports.printRoutes = (routes) => {
  if(!routes) {
    console.log('Cannot find any routes! :(')
  }
  const {printRouteInformation, printLeg} = exports
  routes.forEach(route => {
    console.log('')
    printRouteInformation(route)
    route.legs.forEach(leg => printLeg(route, leg))
    console.log('')
  })
}