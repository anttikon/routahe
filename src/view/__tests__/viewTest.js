import {printSearchInfo} from '../view'
import sinon from 'sinon'

describe('view', function() {
  it('printSearchInfo', () => {
    sinon.spy(console, 'log')
    printSearchInfo({label: 'Jäkälä'}, {label: 'Öylätti'})
    expect(console.log.callCount).toBe(1)
    expect(console.log.getCall(0).args).toMatchSnapshot()
  })
})