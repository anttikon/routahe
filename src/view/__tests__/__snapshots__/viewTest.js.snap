exports[`view printSearchInfo 1`] = `
Array [
  "[1mJ√§k√§l√§[22m - [1m√ñyl√§tti[22m",
]
`;
